tests/lapi/coroutine_test.lua

```diff
@@ -0,0 +1,177 @@
+
+Values held in open upvalues of suspended threads may be
+incorrectly collected,
+https://www.lua.org/bugs.html#5.0.2-3
+
+Attempt to resume a running coroutine crashes Lua,
+https://www.lua.org/bugs.html#5.0-2
+
+debug.getlocal on a coroutine suspended in a hook can crash the interpreter,
+https://www.lua.org/bugs.html#5.3.0-2
+
+debug.sethook/gethook may overflow the thread's stack,
+https://www.lua.org/bugs.html#5.1.2-13
+
+Memory hoarding when creating Lua hooks for coroutines,
+https://www.lua.org/bugs.html#5.2.0-1
+
+Synopsis:
+
+coroutine.close(co)
+coroutine.create(f)
+coroutine.isyieldable([co])
+coroutine.resume(co [, val1, ...])
+coroutine.running()
+coroutine.status(co)
+coroutine.wrap(f)
+coroutine.yield(...)
+]]
+
+local luzer = require("luzer")
+local test_lib = require("lib")
+
+local CORO_OBJECTS
+
+-- Possible coroutine statuses, described in Lua 5.1 Reference Manual,
+-- https://www.lua.org/manual/5.4/manual.html#6.2
+local CORO_STATUS = {
+    DEAD = "dead",
+    NORMAL = "normal",
+    RUNNING = "running",
+    SUSPENDED = "suspended",
+}
+
+local CORO_ACTION_NAME = {
+    "close",
+    "create",
+    "resume",
+    "yield",
+}
+
+  -- Forward declaration.
+local coro_function
+
+local function hook_func(_event)
+    -- Accessing Locals,
+    -- https://www.lua.org/pil/23.1.1.html.
+    local level = 2
+    local i = 1
+    while true do
+        local name, _ = debug.getlocal(level, i)
+        if not name then break end
+        i = i + 1
+    end
+    -- Accessing Upvalues,
+    -- https://www.lua.org/pil/23.1.2.html
+    local func = debug.getinfo(level).func
+    i = 1
+    while true do
+        local name, _ = debug.getupvalue(func, i)
+        if not name then break end
+        i = i + 1
+    end
+end
+
+local function sethook(co, fdp)
+    local set_hook = fdp:consume_boolean()
+    local hook_args = {}
+    if not set_hook then
+        return
+    end
+    table.insert(hook_args, hook_func)
+    table.insert(hook_args, fdp:oneof({"c", "r", "l"}))
+    debug.sethook(co, unpack(hook_args))
+end
+
+local function coro_random_action(fdp, coro_max_number)
+    local action = fdp:oneof(CORO_ACTION_NAME)
+    if action == "create" or #CORO_OBJECTS < coro_max_number then
+        local co = coroutine.create(coro_function)
+        table.insert(CORO_OBJECTS, co)
+    end
+
+    action = fdp:oneof(CORO_ACTION_NAME)
+    local co, co_idx = fdp:oneof(CORO_OBJECTS)
+    if coroutine.status(co) == CORO_STATUS["DEAD"] then
+        assert(co_idx)
+        table.remove(CORO_OBJECTS, co_idx)
+        return
+    elseif action == "close" then
+        coroutine.close(co)
+    elseif action == "yield" and coroutine.isyieldable(co) then
+        coroutine.yield(co)
+    elseif action == "resume" then
+        coroutine.resume(co)
+    end
+    sethook(co, fdp)
+end
+
+coro_function = function(fdp, coro_max_number)
+    local iter = fdp:consume_integer(1, test_lib.MAX_INT)
+    for _ = 1, iter do
```

Comment:
Maybe it is better to limit the amount of these mutations? `MAX_INT` looks like too match.

Reply:
Reduced to 1000, is it ok?

```diff
@@ -154,7 +160,8 @@ local function coro_random_action(fdp, coro_max_number)
 end
 
 coro_function = function(fdp, coro_max_number)
-    local iter = fdp:consume_integer(1, test_lib.MAX_INT)
+    local MAX_N = 1000
+    local iter = fdp:consume_integer(1, MAX_N)
     for _ = 1, iter do
         coro_random_action(fdp, coro_max_number)
     end
```